staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+		// glog.V(2).Infof("Error marshaling secure server info: %v", err)
+		glog.Infof("Error marshaling secure server info: %v", err)
+	}
+}
+
+func NoSecureHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "{ \"Error\": \"No secure server configured\" }")
+}
+
+// tcpConnectionChainListener is responsible to invoking any connection handling
+// code which is invoked after AcceptTCP() has been called but before the
+// connection is usable by the server for communicating with the client. We
+// assume if an error is returned from a chain handler that it will return a nil
+// conn and have already closed that conn. Init() is used to allow low levels tweaks
+// on the TCPConn.
+type tcpConnectionChainHandler interface {
+	Init(*net.TCPConn) error
+}
+
+type tcpConnectionChainListener struct {
+	*net.TCPListener
+	chain []tcpConnectionChainHandler
+}
+
+func (l tcpConnectionChainListener) Accept() (net.Conn, error) {
+	tc, err := l.AcceptTCP()
+	if err != nil {
+		return nil, err
+	}
+	for _, handler := range l.chain {
+		err = handler.Init(tc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tc, nil
+}
+
 // tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
 //
 // Copied from Go 1.7.2 net/http/server.go
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type tcpKeepAliveHandler struct{}
+
+func (ln tcpKeepAliveHandler) Init(conn *net.TCPConn) error {
+	conn.SetKeepAlive(true)
+	conn.SetKeepAlivePeriod(defaultKeepAlivePeriod)
+	return nil
+}
+
+// ipBasedLimitHandler accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+
+type ipBasedLimit struct {
+	// The maximum concurent connections an IP is allowed to have open
+	maxPerIP int
+
+	// The total maximum concurrent connections allowed open except those from localhost
+	max int
+
+	// Lock used to ensure the limits and total are updated atomicaly
+	mutex sync.Mutex
+
+	// Map in which we map the serialized ip to the current count
+	limits map[string]int
+
+	// Map in which we retain the lookup from connection to decrementIP call
+	decrementers map[string]func()
+
+	// Current total count of the connections in existance except those from localhost
+	total int
+}
+
+func (i *ipBasedLimit) incrementIP(conn, sourceIP string) error {
+	if i.max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(sourceIP)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", sourceIP)
+		return IPLimitExceededError{ip: sourceIP, reason: "unparseable ip address"}
```

Comment:
Assuming you are referring to the error, the ip will be added when the actual Error() message is generated. The full Error() message should look like 
"Accept() blocked connection request from 1.2.3.4 because unparseable ip address."